### PR TITLE
Fixes bug in Pattern::Selection #3927

### PR DESCRIPTION
**Pull request by sairajat**
fixes https://github.com/dealii/dealii/issues/3927

### CODE REVIEW

**Comment by sairajat** [comment]
This fixes #3927  Also, I am yet to make changes to changes.h

**Comment by bangerth** [comment]
Yes, this looks great. The next step would be to come up with a test. The way this works is that you would copy one of the tests in `tests/parameter_handler`, let it run on some input that has the offending whitespace, and make sure that the output you want.

**Comment by sairajat** [comment]
@bangerth  How do I run the tests ?

Getting this error on running ```cmake .``` in the ```tests/parameter_handler``` directory.

```
CMake Error at /home/rajat/Documents/Code-Libraries/deal/dealii8.5/dealii-8.5.0/tests/setup_testsubproject.cmake:1 (FIND_PACKAGE):
  Could not find a package configuration file provided by "deal.II"
  (requested version 8.5.0) with any of the following names:

    deal.IIConfig.cmake
    deal.ii-config.cmake

  Add the installation prefix of "deal.II" to CMAKE_PREFIX_PATH or set
  "deal.II_DIR" to a directory containing one of the above files.  If
  "deal.II" provides a separate development package or SDK, be sure it has
  been installed.
Call Stack (most recent call first):
  CMakeLists.txt:2 (INCLUDE)


-- Configuring incomplete, errors occurred!

```

**Comment by sairajat** [comment]
Hello Professor,

I figured out how to run tests. I ran the tests. I got this result.
```
83% tests passed, 20 tests failed out of 119

Total Test time (real) =  64.33 sec

The following tests FAILED:
	 13 - parameter_handler/parameter_handler_read_xml_error_04.debug (Failed)
	 14 - parameter_handler/parameter_handler_read_xml_error_04.release (Failed)
	 23 - parameter_handler/parameter_handler_11.debug (Failed)
	 24 - parameter_handler/parameter_handler_11.release (Failed)
	 25 - parameter_handler/parameter_handler_read_xml_error_03.debug (Failed)
	 26 - parameter_handler/parameter_handler_read_xml_error_03.release (Failed)
	 63 - parameter_handler/parameter_handler_read_xml_error_02.debug (Failed)
	 64 - parameter_handler/parameter_handler_read_xml_error_02.release (Failed)
	 67 - parameter_handler/parameter_handler_backslash_05.debug (Failed)
	 68 - parameter_handler/parameter_handler_backslash_05.release (Failed)
	 71 - parameter_handler/parameter_handler_read_xml_error_01.debug (Failed)
	 72 - parameter_handler/parameter_handler_read_xml_error_01.release (Failed)
	 96 - parameter_handler/parameter_handler_backslash_03.debug (Failed)
	 97 - parameter_handler/parameter_handler_backslash_03.release (Failed)
	 98 - parameter_handler/parameter_handler_10.debug (Failed)
	 99 - parameter_handler/parameter_handler_10.release (Failed)
	102 - parameter_handler/parameter_handler_backslash_04.debug (Failed)
	103 - parameter_handler/parameter_handler_backslash_04.release (Failed)
	106 - parameter_handler/parameter_handler_1_include_fail.debug (Failed)
	107 - parameter_handler/parameter_handler_1_include_fail.release (Failed)
Errors while running CTest
```

**Comment by bangerth** [comment]
Well, that looks disturbing with all of the tests failing :-) Take a look here for how to set up and run the testsuite: http://dealii.org/developer/developers/testsuite.html

**Comment by sairajat** [comment]
@bangerth  I followed the instructions but I think I am still doing something wrong as I dont see the parameter_handler_3.cc in the tests directory being run. 

In the build directory, I ran ```make setup_tests```. In the ```build/tests``` I ran ```ctest -j8 -R "parameter_handler"```. This is the output I got. 
```
    Start 2454: all-headers/base/parameter_handler.h.debug
    Start 2455: all-headers/base/parameter_handler.h.release
    Start 4689: serialization/parameter_handler.debug
    Start 4690: serialization/parameter_handler.release
1/4 Test #2454: all-headers/base/parameter_handler.h.debug .....   Passed    2.83 sec
2/4 Test #2455: all-headers/base/parameter_handler.h.release ...   Passed    2.83 sec
3/4 Test #4690: serialization/parameter_handler.release ........   Passed    4.50 sec
4/4 Test #4689: serialization/parameter_handler.debug ..........   Passed    4.50 sec

100% tests passed, 0 tests failed out of 4

Total Test time (real) =   4.97 sec
```

**Comment by bangerth** [comment]
This seems to exclude all tests that are actually located in `tests/parameter_handler`. What happens if you execute the command in `build/`, rather than `build/tests/`?

**Comment by bangerth** [comment]
And did you set `TEST_PICKUP_REGEX` to anything before calling `make setup_tests`?

**Comment by sairajat** [comment]
@bangerth  Yes, I tried with TEST_PICKUP_REGEX="paramater_handler" once.

**Comment by sairajat** [comment]
Doing ``` ctest -j8 -R "tests/parameter_handler" ``` in the build directory gives

```No tests were found!!!```

**Comment by bangerth** [comment]
Drop the `tests/` and just say `ctest -R parameter_handler`. That ought to be good enough.

If it isn't, try
```
  unset TEST_PICKUP_REGEX
  make setup_tests
  ctest -R parameter_handler
```

**Comment by sairajat** [comment]
Doing ```ctest -R parameter_handler``` in the build directory gives the same thing.
```
    Start 2454: all-headers/base/parameter_handler.h.debug
1/4 Test #2454: all-headers/base/parameter_handler.h.debug .....   Passed    1.75 sec
    Start 2455: all-headers/base/parameter_handler.h.release
2/4 Test #2455: all-headers/base/parameter_handler.h.release ...   Passed    1.77 sec
    Start 4689: serialization/parameter_handler.debug
3/4 Test #4689: serialization/parameter_handler.debug ..........   Passed    0.13 sec
    Start 4690: serialization/parameter_handler.release
4/4 Test #4690: serialization/parameter_handler.release ........   Passed    0.12 sec

100% tests passed, 0 tests failed out of 4

```

Doing the other option also gives the same thing.
```  
unset TEST_PICKUP_REGEX
  make setup_tests
  ctest -R parameter_handler
```

```

    Start 2454: all-headers/base/parameter_handler.h.debug
1/4 Test #2454: all-headers/base/parameter_handler.h.debug .....   Passed    1.77 sec
    Start 2455: all-headers/base/parameter_handler.h.release
2/4 Test #2455: all-headers/base/parameter_handler.h.release ...   Passed    1.80 sec
    Start 4689: serialization/parameter_handler.debug
3/4 Test #4689: serialization/parameter_handler.debug ..........   Passed    0.12 sec
    Start 4690: serialization/parameter_handler.release
4/4 Test #4690: serialization/parameter_handler.release ........   Passed    0.12 sec

100% tests passed, 0 tests failed out of 4

Total Test time (real) =   4.20 sec
```

**Comment by bangerth** [comment]
Hm, that is bizarre. If I do these steps, it runs 123 tests.

In your source directory, what do you have in `tests/parameter_handler`?

**Comment by sairajat** [comment]
@bangerth  Sir, I was able to run the tests (yes, all 123).  I was able to do that by replacing the directory ```tests/parameter_handler``` from the github source again.

Also, In the ```/tests/parameter_handler/parameter_handler_3.cc```,  If I put an extra space before a, it fails with the old source code, but passes with the file with bug-fixed. So, should an entirely new test be created  or just adding a space is fine or should a new  call be made to ```declare_entry()``` with the old arguments but with added space.

``` prm.declare_entry ("string list",
                         "a",
                         Patterns::List(Patterns::Selection(" a|b|c|d|e|f|g|h")),
                         "docs 1");
```

**Comment by bangerth** [comment]
Yes, just copy the entire test to, say, `parameter_handler_3_with_space.cc` or so, add the space, and copy the old `.output` file to the new test's name (the output file should be exactly the same).

Then run `make setup_tests` again and it should pick up your test.

**Comment by sairajat** [comment]
@bangerth  Is this all right ?

File: tests/parameter_handler/parameter_handler_3_with_space.cc [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+
+
+// test the output generated by ParameterHandler::print_parameters(Text)
```
Comment (bangerth) [inline tests/parameter_handler/parameter_handler_3_with_space.cc]: Can you change this to say
```
  Like parameter_handler_03, but use a MultipleSelection pattern that starts with a space; eat that space
```

**Comment by bangerth** [comment]
Yes, close!

/run-tests

**Comment by sairajat** [comment]
@bangerth  Sir, kindly let me know if this is fine. I will then add changes in "changes.h".

**Comment by bangerth** [comment]
Yes, great. Please just put a file into `doc/news/changes/minor` in the same format as the other files there.

It would also be great if you could squash all of your commits into one if you know how to do that. Alternatively, into one for the patch to `parameter_handler.cc`, one for the new testcase, and one commit for the changelog entry.

**Comment by davydden** [comment]
@sairajat  if you don't know how to squash, we can do it ourselves during merging.

**Comment by sairajat** [comment]
@davydden  Thanks. Yes, I don't know how to squash. I am still learning how to use git. I will make a final commit with the file in ```doc/news/changes/minor```. 

I will learn squash and some other advanced git features definitely by next time  :)

File: tests/parameter_handler/parameter_handler_3_with_space.cc [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    {
+      std::ofstream logfile("output");
+      deallog.attach(logfile);
+      deallog.threshold_double(1.e-10);
+
+      ParameterHandler prm;
+      prm.enter_subsection ("Testing");
+      prm.declare_entry ("string list",
+                         "a",
+                         Patterns::List(Patterns::Selection(" a|b|c|d|e|f|g|h")),
```
Comment (davydden) [inline tests/parameter_handler/parameter_handler_3_with_space.cc]: could you please add another ` prm.declare_entry` with white space at the end? That's not tested, as far as I can tell.

**Comment by davydden** [comment]
@sairajat btw, you can make github automatically close issues which are being fixed by a given PR. I edited your OP to do that.

**Comment by sairajat** [comment]
@davydden  Please let me know if this is fine. Thanks.

File: tests/parameter_handler/parameter_handler_3_with_space.output [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+# Listing of Parameters
+# ---------------------
+subsection Testing
+  # docs 3
+  set double      = 3.1415926
+  set int         = 3         # default: 1
+
+  # docs 1
+  set string list1 = a, b, c   # default: a
+  #docs 2
```
Comment (davydden) [inline tests/parameter_handler/parameter_handler_3_with_space.output]: does the test pass? There is white space in `docs 1` and no whitespace here. I wonder why.

**Comment by bangerth** [comment]
Yes, looks good if @sairajat  can confirm that the new test really succeeds.

**Comment by sairajat** [comment]
Respected Sir,
You were right. There was something wrong. The tests pass succeed now. 

Also, there are some tests in parameter_handler that are failing on my machine even when I make a fresh installation from the master branch that I have synced 5 days ago (without these changes). 

``` ctest -j8 -R parameter_handler_3_with_space*
Test project /home/rajat/Documents/Code-Libraries/deal/dealii8.5/dealii-8.5.0/build-new
    Start 80: parameter_handler/parameter_handler_3_with_space.release
    Start 79: parameter_handler/parameter_handler_3_with_space.debug
1/2 Test #79: parameter_handler/parameter_handler_3_with_space.debug .....   Passed    0.13 sec
2/2 Test #80: parameter_handler/parameter_handler_3_with_space.release ...   Passed    0.14 sec

100% tests passed, 0 tests failed out of 2

Total Test time (real) =   0.14 sec
```


```
83% tests passed, 20 tests failed out of 121

Total Test time (real) =   3.44 sec

The following tests FAILED:
	 13 - parameter_handler/parameter_handler_read_xml_error_04.debug (Failed)
	 14 - parameter_handler/parameter_handler_read_xml_error_04.release (Failed)
	 23 - parameter_handler/parameter_handler_11.debug (Failed)
	 24 - parameter_handler/parameter_handler_11.release (Failed)
	 25 - parameter_handler/parameter_handler_read_xml_error_03.debug (Failed)
	 26 - parameter_handler/parameter_handler_read_xml_error_03.release (Failed)
	 63 - parameter_handler/parameter_handler_read_xml_error_02.debug (Failed)
	 64 - parameter_handler/parameter_handler_read_xml_error_02.release (Failed)
	 67 - parameter_handler/parameter_handler_backslash_05.debug (Failed)
	 68 - parameter_handler/parameter_handler_backslash_05.release (Failed)
	 71 - parameter_handler/parameter_handler_read_xml_error_01.debug (Failed)
	 72 - parameter_handler/parameter_handler_read_xml_error_01.release (Failed)
	 98 - parameter_handler/parameter_handler_backslash_03.debug (Failed)
	 99 - parameter_handler/parameter_handler_backslash_03.release (Failed)
	100 - parameter_handler/parameter_handler_10.debug (Failed)
	101 - parameter_handler/parameter_handler_10.release (Failed)
	104 - parameter_handler/parameter_handler_backslash_04.debug (Failed)
	105 - parameter_handler/parameter_handler_backslash_04.release (Failed)
	108 - parameter_handler/parameter_handler_1_include_fail.debug (Failed)
	109 - parameter_handler/parameter_handler_1_include_fail.release (Failed)
Errors while running CTest
```

**Comment by davydden** [comment]
p.s. 

> Also, there are some tests in parameter_handler that are failing on my machine

please open a separate issue for that with details of why things fail (you can run `ctest -V -R "parameter_handler/parameter_handler_read_xml_error_04.debug"` to have verbose output for a given test)

**Comment by davydden** [comment]
@sairajat wait a second, do those failing test work for your on `master` branch (without those modifications)?

**Comment by sairajat** [comment]
Respected Sir,

These tests fail with and without these modifications. I am posting the output below for running the tests without the modification. You can see that the test ```parameteter_handler_3_with_space.cc``` also fails as it needs modification.

```
82% tests passed, 22 tests failed out of 121

Total Test time (real) =  45.91 sec

The following tests FAILED:
	 13 - parameter_handler/parameter_handler_read_xml_error_04.debug (Failed)
	 14 - parameter_handler/parameter_handler_read_xml_error_04.release (Failed)
	 23 - parameter_handler/parameter_handler_11.debug (Failed)
	 24 - parameter_handler/parameter_handler_11.release (Failed)
	 25 - parameter_handler/parameter_handler_read_xml_error_03.debug (Failed)
	 26 - parameter_handler/parameter_handler_read_xml_error_03.release (Failed)
	 63 - parameter_handler/parameter_handler_read_xml_error_02.debug (Failed)
	 64 - parameter_handler/parameter_handler_read_xml_error_02.release (Failed)
	 67 - parameter_handler/parameter_handler_backslash_05.debug (Failed)
	 68 - parameter_handler/parameter_handler_backslash_05.release (Failed)
	 71 - parameter_handler/parameter_handler_read_xml_error_01.debug (Failed)
	 72 - parameter_handler/parameter_handler_read_xml_error_01.release (Failed)
	 79 - parameter_handler/parameter_handler_3_with_space.debug (Failed)
	 80 - parameter_handler/parameter_handler_3_with_space.release (Failed)
	 98 - parameter_handler/parameter_handler_backslash_03.debug (Failed)
	 99 - parameter_handler/parameter_handler_backslash_03.release (Failed)
	100 - parameter_handler/parameter_handler_10.debug (Failed)
	101 - parameter_handler/parameter_handler_10.release (Failed)
	104 - parameter_handler/parameter_handler_backslash_04.debug (Failed)
	105 - parameter_handler/parameter_handler_backslash_04.release (Failed)
	108 - parameter_handler/parameter_handler_1_include_fail.debug (Failed)
	109 - parameter_handler/parameter_handler_1_include_fail.release (Failed)
Errors while running CTest
```

**Comment by davydden** [comment]
> You can see that the test parameteter_handler_3_with_space.cc also fails as it needs modification.

Oh, I see that you did NOT create a separate branch for this fix/PR. Normally you would keep `master` branch in your fork untouched and create a separate branch for a new feature or a bugfix.

So what I mean is to see if those tests also fail without changes you introduce in this PR. In your case you would need to do something along the lines:
```
# this will checkout branch "master" in remote "origin"
$ git checkout origin/master
# compile dealii
$ make all -j8
$ make setup_tests_parameter_handler
$ ctest -R "parameter_handler"
```
if the tests fail indeed, please open a dedicate issue here https://github.com/dealii/dealii/issues
I don't expect this PR to have an effect on those failing tests, though.

**Comment by sairajat** [comment]
@davydden  Sir,
I did the same thing. I commented out the modifications I introduced in the ```parameter_handler.cc``` file. I recompiled deal.ii and run the tests. This why  ```parameter_handler_3_with_space.cc``` also failed as it needed the lines I commented. 

Here, I assumed that the presence of a test (```parameter_handler_3_with_space.cc ``` in this case) will not affect other tests.

**Comment by davydden** [comment]
> Here, I assumed that the presence of a test (parameter_handler_3_with_space.cc in this case) will not affect other tests.

that's correct.

**Comment by sairajat** [comment]
I will open a new issue dedicated to tests failing.

**Comment by masterleinad** [comment]
At least for me, this PR passes all `parameter`tests.

**Comment by drwells** [comment]
Before this is merged, would you mind rewriting your commit history a bit? This is just one fix and it should be described in the log by one commit.

**Comment by sairajat** [comment]
@drwells  Sir, are you talking about squashing commits ? I am new to git and I don't know how to do that.  Davyden said that the squashing can be done by him / other dealii authors before merging. 

If there is something else that you need me to do, I will be happy to do it. 
Thanks.

**Comment by drwells** [comment]
You are right. @davydden did volunteer to do that so we should be set :)

**Comment by bangerth** [comment]
Thanks, @sairajat , for sticking with us through this whole process! This was a lot of back-and-forth, but it will get easier in the future as you learn what to look out for when writing patches. I hope you feel like you learned something through all of these comments and requests, and I hope this won't be your last patch!

**Comment by sairajat** [comment]
@bangerth  Respected Sir, actually I would like to thank you, and @davydden  for helping me out during this. It was a small bug but I definitely learned a lot during this whole process.  I will keep reporting and  contributing while using deal.ii. :)

**Comment by bangerth** [comment]
@davydden deserves the credit. I was only doing my job :-)

Anyway, glad to hear that you thought it was a worthwhile experience and that you may want to do it again!

**Pull request merged**